src/audio/drc/drc_generic.c

```diff
@@ -0,0 +1,435 @@
+
+	if (is_releasing) {
+		/* Release mode - compression_diff_db should be negative dB */
+		state->max_attack_compression_diff_db = INT32_MIN;
+
+		/* Fix gremlins. */
+		if (isbadf(compression_diff_db))
+			compression_diff_db = -1;
+
+		/* Adaptive release - higher compression (lower
+		 * compression_diff_db) releases faster. Contain within range:
+		 * -12 -> 0 then scale to go from 0 -> 3
+		 */
+		float x = compression_diff_db;
+		x = max(-12.0f, x);
+		x = min(0.0f, x);
+		x = 0.25f * (x + 12);
+
+		/* Compute adaptive release curve using 4th order polynomial.
+		 * Normal values for the polynomial coefficients would create a
+		 * monotonically increasing function.
+		 */
+		float x2 = x * x;
+		float x3 = x2 * x;
+		float x4 = x2 * x2;
+		float release_frames = kA + kB * x + kC * x2 + kD * x3 + kE * x4;
+
+#define kSpacingDb 5
+		float db_per_frame = kSpacingDb / release_frames;
+		envelope_rate = decibels_to_linear(db_per_frame);
+	} else {
+		/* Attack mode - compression_diff_db should be positive dB */
+
+		/* Fix gremlins. */
+		if (isbadf(compression_diff_db))
+			compression_diff_db = 1;
+
+		/* As long as we're still in attack mode, use a rate based off
+		 * the largest compression_diff_db we've encountered so far.
+		 */
+		state->max_attack_compression_diff_db =
+			max(state->max_attack_compression_diff_db,
+			    Q_CONVERT_FLOAT(compression_diff_db, 20));
+
+		float eff_atten_diff_db =
+			max(0.5f, Q_CONVERT_QTOF(state->max_attack_compression_diff_db, 20));
+
+		float x = 0.25f / eff_atten_diff_db;
+		envelope_rate = 1 - powf(x, 1 / attack_frames);
+	}
+
+	state->envelope_rate = Q_CONVERT_FLOAT(envelope_rate, 30);
+	state->scaled_desired_gain = Q_CONVERT_FLOAT(scaled_desired_gain, 30);
+}
+
+/* Calculate compress_gain from the envelope and apply total_gain to compress
+ * the next output division. */
+static void drc_compress_output(struct drc_state *state, struct sof_drc_params *p, int nch)
+{
+	const float master_linear_gain = Q_CONVERT_QTOF(p->master_linear_gain, 30);
+	const float envelope_rate = Q_CONVERT_QTOF(state->envelope_rate, 30);
+	const float scaled_desired_gain = Q_CONVERT_QTOF(state->scaled_desired_gain, 30);
+	const float compressor_gain = Q_CONVERT_QTOF(state->compressor_gain, 30);
+	const int div_start = state->pre_delay_read_index;
+	int count = DRC_DIVISION_FRAMES / 4;
+
+	int i, j, ch, inc;
+	float sample;
+
+	/* Exponential approach to desired gain. */
+	if (envelope_rate < 1) {
+		/* Attack - reduce gain to desired. */
+		float c = compressor_gain - scaled_desired_gain;
+		float base = scaled_desired_gain;
+		float r = 1 - envelope_rate;
+		float x[4] = { c * r, c * r * r, c * r * r * r,
+			       c * r * r * r * r };
+		float r4 = r * r * r * r;
+
+		i = 0;
+		inc = 0;
+		while (1) {
+			for (j = 0; j < 4; j++) {
+				/* Warp pre-compression gain to smooth out sharp
+				 * exponential transition points.
+				 */
+				float post_warp_compressor_gain = warp_sinf(x[j] + base);
```

Comment:
sinf() looks expensive computationally. Is this executed when processing copy()?

Reply:
I think it is called from copy

Reply:
Unfortunately all math functions in the latest code are called from copy, and will be called multiple times.
We will try to implement the approximated cheaper alternatives.